File: 205.isomorphic-strings.java
/*
 * @lc app=leetcode id=205 lang=java
 *
 * [205] Isomorphic Strings
 */
class Solution {
    public boolean isIsomorphic(String s, String t) {
        if (s == null || t == null) return false;
        char[] ca = t.toCharArray();
        if (ca.length != s.length()) return false;
        Map<Character, Character> dict = new HashMap<>();
        Set<Character> seem = new HashSet<>();
        for (int i = 0; i < s.length(); ++i) {
            Character c = dict.get(s.charAt(i));
            if (c == null) {
                dict.put(s.charAt(i), ca[i]);
                if (seem.contains(ca[i])) return false;
                seem.add(ca[i]);
            }
            else if (c != ca[i]) return false;
        }
        return true;
    }
}

File: go/205.isomorphic-strings.go
/*
 * @lc app=leetcode id=205 lang=golang
 *
 * [205] Isomorphic Strings
 */

// @lc code=start

func isIsomorphic(s string, t string) bool {
	return isInjective(&s, &t) && isInjective(&t, &s)
}

func isInjective(ps *string, pt *string) bool {
	s, t := *ps, *pt
	mapping := make(map[byte]byte)
	for i := 0; i < len(s); i++ {
		if v, b := mapping[t[i]]; b {
			if v != s[i] {
				return false
			}
		} else {
			mapping[t[i]] = s[i]
		}
	}
	return true
}

// @lc code=end

